toreview/BitSetUtils.java

```diff
@@ -0,0 +1,174 @@
+package com.gruppo1.distributednetworkmanager;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.BitSet;
+
+public class BitSetUtils {
+
+    private static int minLength = 64;
+    private static final String NOT_VALID_NUMBIT_EXCEPTION_MSG = "numBits isn't > 0 or a multiple of 64";
+    private static final String NOT_VALID_NUMBIT_SHA_EXCEPTION_MSG = "numBits isn't > 0 or <= 160";
+    private static final String INVALID_HEX_CHAR_MSG = "Invalid Hexadecimal Character: ";
+    private static final String INVALID_HEX_STRING_MSG = "Invalid hexadecimal String supplied.";
+    private static final String UTILS_TAG = BitSetUtils.class.toString();
+    private static final String SHA_1 = "SHA-1";
+
+    /**
+     * @param first the first binary key
+     * @param second the second binary key
+     * @return The distance of the Keys, calculated in XOR logic
+     */
+    public static BitSet distanceFrom(@NonNull BitSet first, @NonNull BitSet second) {
+        BitSet distance = (BitSet) first.clone();
+        distance.xor(second);
+        return distance;
+    }
+
+    /**
+     * @param toHash  the String (a key) of which generate hashcode
+     * @param numBits number of bits that the hash code will be constituted of, must be multiple of 64 and >0
+     * @return the bitSet containing the hash of the peer's address, bitSet's length is a multiple of 64
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't multiple of 64
+     */
+    public static BitSet hash(@NonNull String toHash, int numBits) {
+        if (numBits > 0 && numBits % minLength == 0) {
+            int numLong = numBits / minLength;
+            long[] numbers = new long[numLong];
+            for (int i = 0; i < numLong; i++) {
+                numbers[numLong - (i + 1)] = Double.doubleToLongBits(hash(toHash) / (i + 1));
+            }
+            return BitSet.valueOf(numbers);
+        } else throw new IllegalArgumentException(NOT_VALID_NUMBIT_EXCEPTION_MSG);
+    }
+
+    /**
+     * @param toHash  the String (a key) of which generate hashcode
+     * @param numBits number of bits that the hash code will be constituted of, must be > 0 &&  <= 160
+     * @return the bitSet containing the hash (SHA-1) of the bytes in input, truncated to numBits, bitSet's size is <= numBits
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't in its bounds
+     */
+    public static BitSet hash(@NonNull byte[] toHash, int numBits) {
+        if (numBits > 0 && numBits <= 160) {
+            byte[] digest = {0};
+            try {
+                MessageDigest md = MessageDigest.getInstance(SHA_1);
+                md.update(toHash);
+                digest = md.digest();

```

Comment:
Potevate condensare queste due righe in `md.digest(toHash)`.

Reply:
In effetti è vero.